### PR TITLE
Dev container

### PR DESCRIPTION
**Pull request by okutter**
Utility build script and Dockerfile for building Holohub development container and launching it.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ARG BASE_IMAGE
+ARG GPU_TYPE
+
+FROM ${BASE_IMAGE} as base
+
+ARG DEBIAN_FRONTEND=noninteractive
+
+RUN apt update && \ 
+    apt install --no-install-recommends -y \
+    libv4l-dev=1.18.0-2build1
```
Comment (jjomier) [inline Dockerfile:32]: @okutter Can we move this dependency in the run setup instead?

File: README.md [hunk 2]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
 HoloHub has been tested and is known to run on Ubuntu 20.04. Other versions of Ubuntu or OS distributions may result in build and/or runtime issues.
 
-1. Refer to the [Holoscan SDK README](https://github.com/nvidia-holoscan/holoscan-sdk/blob/main/README.md) for ways to install Holoscan SDK: Debian package, Python wheels, container or from source.
 
-2. Clone this repository.
+1. Clone this repository.
+
+2. Choose to build Holohub using development container or using a local environment.
+
+**Container build**
```
Comment (wendell-hom) [inline README.md:21]:
```suggestion
### Container build
```

File: README.md [hunk 3]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+
+***Build container:*** Run the following command from the holohub directory to build the development container.
+
+```bash
+  ./dev_container build
+```
+
+***Note:*** The development container script ```dev_container``` will by default use [NGC's Holoscan SDK container](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan) for the local GPU configuration by detecting if the system is using an iGPU (integrated GPU) or a dGPU (discrete GPU). 
+
+For example on x86 ```docker images``` will list the following new images after ```./dev_container build``` has completed.
```
Comment (wendell-hom) [inline README.md:31]:
```suggestion
For example, on a system with dGPU ```docker images``` will list the following new images after ```./dev_container build``` has completed.
```

File: README.md [hunk 4]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+
+
+***Notes for AGX Orin Dev Kit***: 
+
+(1) On AGX Orin Dev Kit the launch script will add ```--privileged``` and ```--group-add video``` to the docker run command for the Holohub sample apps to work. Please also make sure that the current user is member of the group video.  
+
+(2) When building Holoscan SDK on AGX Orin Dev Kit from source please add the option  ```--cudaarchs all``` to the ```./run build``` command to include support for AGX Orin's iGPU.
+
+
+**Local build**
```
Comment (wendell-hom) [inline README.md:137]:
```suggestion
### Local build
```

File: README.md [hunk 5]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+
+1. Custom base image
+
+It is possible to configure a custom base image for building the Holohub container. E.g., if you built Holoscan SDK locally and want to use the locally built container as base image, use the following command: 
+
+```bash
+  ./dev_container build --base_img holoscan-sdk-dev:latest --img_fqn holohub:sdk-dev-latest
+```
+
+where ```--base_img```  is used to configure the base container image and ```--img_fqn``` to define the fully qualified image name.
```
Comment (wendell-hom) [inline README.md:49]:
```suggestion
where ```--base_img```  is used to configure the base container image and ```--img``` to define the fully qualified image name. 
```
Probably don't need 'fqn' in the option name

File: README.md [hunk 6]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+```
+
+***Advanced build options***
+
+1. Custom base image
+
+It is possible to configure a custom base image for building the Holohub container. E.g., if you built Holoscan SDK locally and want to use the locally built container as base image, use the following command: 
+
+```bash
+  ./dev_container build --base_img holoscan-sdk-dev:latest --img_fqn holohub:sdk-dev-latest
```
Comment (wendell-hom) [inline README.md:46]:
```suggestion
  ./dev_container build --base_img holoscan-sdk-dev:latest --img holohub:sdk-dev-latest
```
Probably don't need to add 'fqn' to the option name

File: README.md [hunk 7]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+user@ubuntu-20-04:/media/data/github/holohub$ docker images
+REPOSITORY                               TAG               IMAGE ID       CREATED          SIZE
+holohub                                  sdk-dev-latest    cb0231f77856   54 seconds ago   8.22GB
+```
+
+
+2. Using custom Dockerfile
+
+```bash
+  ./dev_container build  --docker_file PATH_TO_DOCKERFILE  --img_fqn holohub-debug:local-sdk-v0.5.1
```
Comment (wendell-hom) [inline README.md:63]:
```suggestion
  ./dev_container build  --docker_file <path_to_dockerfile>  --img holohub-debug:local-sdk-v0.5.1
```
Use similar style as other parts of holohub docs

File: README.md [hunk 8]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+```
+
+where ```--docker_file```  is the path to the container's Dockerfile and ```--img_fqn``` to define the fully qualified image name.
```
Comment (wendell-hom) [inline README.md:66]:
```suggestion
where ```--docker_file```  is the path to the container's Dockerfile and ```--img``` to define the fully qualified image name. 
```

File: README.md [hunk 9]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
-1. Refer to the [Holoscan SDK README](https://github.com/nvidia-holoscan/holoscan-sdk/blob/main/README.md) for ways to install Holoscan SDK: Debian package, Python wheels, container or from source.
 
-2. Clone this repository.
+1. Clone this repository.
+
+2. Choose to build Holohub using development container or using a local environment.
+
+**Container build**
+
+***Build container:*** Run the following command from the holohub directory to build the development container.
```
Comment (wendell-hom) [inline README.md:23]:
```suggestion
***Building dev container:*** Run the following command from the holohub directory to build the development container.
```

File: README.md [hunk 10]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+Build (HOLOHUB_ROOT:/media/data/github/holohub)...
+Build (gpu_type_type:dgpu)...
+Build (base_img:nvcr.io/nvidia/clara-holoscan/holoscan:v0.5.1-dgpu)...
+Build (docker_file_path:/media/data/github/holohub/Dockerfile)...
+Build (img_fqn:holohub:ngc-v0.5.1-dgpu)...
+....
+```
+
+
+***Launch container:***  Run the following command from the holohub directory to launch the default development container built using Holoscan SDK's container from ngc for the local GPU.
```
Comment (wendell-hom) [inline README.md:83]:
```suggestion
***Launching dev container:***  Run the following command from the holohub directory to launch the default development container built using Holoscan SDK's container from ngc for the local GPU.
```

File: README.md [hunk 11]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+```
+
+The launch script will also inspect for available video devices (V4L2, AJA capture boards, Deltacast capture boards) and presence of Deltacast's Videomaster SDK and map it into the development container.
+
+***Advanced launch options***
+
+1. For Holohub image built with locally built Holoscan SDK container
+
+```bash
+  ./dev_container launch --img holohub:local-sdk-latest --local_sdk_root PATH_TO_HOLOSCAN_SDK
```
Comment (wendell-hom) [inline README.md:96]:
```suggestion
  ./dev_container launch --img holohub:local-sdk-latest --local_sdk_root <path_to_holoscan_sdk>
```

File: README.md [hunk 12]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+```
+
+
+***Launch container:***  Run the following command from the holohub directory to launch the default development container built using Holoscan SDK's container from ngc for the local GPU.
+
+```bash
+  ./dev_container launch
+```
+
+The launch script will also inspect for available video devices (V4L2, AJA capture boards, Deltacast capture boards) and presence of Deltacast's Videomaster SDK and map it into the development container.
```
Comment (wendell-hom) [inline README.md:89]:
```suggestion
The launch script will also inspect for available video devices (V4L2, AJA capture boards, Deltacast capture boards) and the presence of Deltacast's Videomaster SDK and map it into the development container.
```

File: README.md [hunk 13]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+
+2. Launch custom Holohub container with fully qualified name, e.g. "holohub:ngc-sdk-sample-app"
+
+```bash
+  ./dev_container launch --img holohub:ngc-sdk-sample-app
+```
+
+***Note***:  To debug the values passed to the docker run command when using add the ```--debug``` option to the launch command.
+
+For example, on x86 with dGPU ```./dev_container launch --debug``` will print the following values.
```
Comment (wendell-hom) [inline README.md:107]:
```suggestion
For example, on an x86_64 system with dGPU ```./dev_container launch --debug``` will print the following values. 
```

File: README.md [hunk 14]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+REPOSITORY                               TAG           IMAGE ID       CREATED         SIZE
+holohub                                  ngc-v0.5.1-dgpu   b6d86bccdcac   9 seconds ago   8.47GB
+nvcr.io/nvidia/clara-holoscan/holoscan   v0.5.1-dgpu       1b4df7733d5b   5 weeks ago     8.04GB
+```
+
+***Advanced build options***
+
+1. Custom base image
+
+It is possible to configure a custom base image for building the Holohub container. E.g., if you built Holoscan SDK locally and want to use the locally built container as base image, use the following command:
```
Comment (wendell-hom) [inline README.md:43]:
```suggestion
It is possible to configure a custom base image for building the Holohub container. E.g., if you built Holoscan SDK locally and want to use the locally built container as the base image, use the following command: 
```

File: README.md [hunk 15]
```diff
@@ -13,16 +13,136 @@ Holohub is a central repository for users and developers of extensions and appli
...
+  ./dev_container launch --img holohub:local-sdk-latest --local_sdk_root PATH_TO_HOLOSCAN_SDK
+```
+
+2. Launch custom Holohub container with fully qualified name, e.g. "holohub:ngc-sdk-sample-app"
+
+```bash
+  ./dev_container launch --img holohub:ngc-sdk-sample-app
+```
+
+***Note***:  To debug the values passed to the docker run command when using add the ```--debug``` option to the launch command.
```
Comment (wendell-hom) [inline README.md:105]:
```suggestion
***Note***:  To print the values passed to the docker run command, add the ```--debug``` option to the launch command.
```

**Comment by wendell-hom** [comment]
@okutter Thank you for adding the dev_container script to holohub!  It looks good, I tried out the dev_container and it looks flexible and pretty easy to use.  The suggestions here are mostly to rephrase some parts of the README.  

Aside from simple rephrasing, 
* @jjomier added a comment about whether we can move the libv4l dependency into `run setup`.  
* for the --img_fqn option, I'm suggesting we can just change that to --img
* for the --debug option flag, I'm wondering if we should update it to --verbose

File: Dockerfile (hunk 1 above)
Comment (okutter) [reply Dockerfile]: @jjomier: Done, see https://github.com/nvidia-holoscan/holohub/commit/a22ca82c38439dcba5957e8029fa5afd2a8ee939

**Review by wendell-hom — APPROVED** [review]
@okutter Looks great!  Thank you!

**Pull request merged**